components/psst/browser/content/psst_tab_helper.cc

```diff
@@ -0,0 +1,136 @@
+#include "base/values.h"
+#include "brave/components/psst/browser/core/psst_rule.h"
+#include "brave/components/psst/browser/core/psst_rule_service.h"
+#include "brave/components/psst/common/features.h"
+#include "components/sessions/content/session_tab_helper.h"
+#include "content/public/browser/browser_context.h"
+#include "content/public/browser/navigation_entry.h"
+#include "content/public/browser/navigation_handle.h"
+#include "content/public/browser/web_contents.h"
+#include "mojo/public/cpp/bindings/associated_remote.h"
+#include "third_party/blink/public/common/associated_interfaces/associated_interface_provider.h"
+
+namespace psst {
+
+// static
+void PsstTabHelper::MaybeCreateForWebContents(content::WebContents* contents,
+                                              const int32_t world_id) {
+  if (contents->GetBrowserContext()->IsOffTheRecord() ||
+      !base::FeatureList::IsEnabled(psst::features::kBravePsst)) {
+    return;
+  }
+
+  psst::PsstTabHelper::CreateForWebContents(contents, world_id);
+}
+
+PsstTabHelper::PsstTabHelper(content::WebContents* web_contents,
+                             const int32_t world_id)
+    : WebContentsObserver(web_contents),
+      content::WebContentsUserData<PsstTabHelper>(*web_contents),
+      world_id_(world_id),
+      psst_rule_service_(PsstRuleService::GetInstance()) {
+  DCHECK(psst_rule_service_);
+}
+
+PsstTabHelper::~PsstTabHelper() = default;
+
+void PsstTabHelper::OnTestScriptResult(
+    std::string policy_script,
+    const content::GlobalRenderFrameHostId render_frame_host_id,
+    const int32_t world_id,
+    base::Value value) {
+  if (value.GetIfBool().value_or(false)) {
+    InsertScriptInPage(render_frame_host_id, world_id, policy_script,
+                       base::DoNothing());
+  }
+}
+
+void PsstTabHelper::InsertTestScript(
+    const content::GlobalRenderFrameHostId render_frame_host_id,
+    const int32_t world_id,
+    MatchedRule rule) {
+  CHECK(render_frame_host_id);
+  InsertScriptInPage(
+      render_frame_host_id, world_id, rule.test_script,
+      base::BindOnce(&PsstTabHelper::OnTestScriptResult,
+                     weak_factory_.GetWeakPtr(), rule.policy_script,
+                     render_frame_host_id, world_id));
+}
+
+void PsstTabHelper::InsertScriptInPage(
+    const content::GlobalRenderFrameHostId render_frame_host_id,
+    const int32_t world_id,
+    const std::string& script,
+    content::RenderFrameHost::JavaScriptResultCallback cb) {
+  content::RenderFrameHost* render_frame_host =
+      content::RenderFrameHost::FromID(render_frame_host_id);
+
+  // Check if render_frame_host is still valid and if starting rfh is the same.
+  if (render_frame_host &&
+      render_frame_host_id ==
+          web_contents()->GetPrimaryMainFrame()->GetGlobalId()) {
+    GetRemote(render_frame_host)
+        ->RequestAsyncExecuteScript(world_id, base::UTF8ToUTF16(script), false,
+                                    true, true, std::move(cb));
```

Comment:
in this case you don't want a result, right? https://github.com/brave/brave-core/pull/20336/files#diff-c6e915dfab9870f210c65819afd77c8b628bc8151c758de6a05cb913c7d9e2e9R57
Maybe make the callback here optional and set to `false` and use `base::DoNothing` if the callback isn't supplied?

Reply:
I'm not sure that adds value over what exists currently.

Reply:
I suppose you could handle this the same way extension code does which is to only pass the callback and not pass `wants_result` here at all. On the renderer side you'd `wants_result` based on `callback.is_null()`